hadml/datamodules/components/herwig.py

```diff
@@ -661,113 +660,296 @@ def get_angles(four_vector):
+        (self._create_data) to get all the hadron PIDs. """
+        all_data = []
 
-    def _create_data(self, line):
+        for raw_path in self.raw_file_list:
+            with open(raw_path) as f:
+                data_list = [self._create_data(line, create_dict=True, n_line=i) 
+                             for i, line in enumerate(f) if line.strip()]
+            print()
+            data_list = np.concatenate(data_list)
+            all_data.append(data_list)
+        
+        all_data = np.array(all_data).flatten()
+        count = Counter(all_data)
+        hadron_pids = list(map(lambda x: x[0], count.most_common()))
+        self.pids_to_ix = {pids: i for i, pids in enumerate(hadron_pids)}
+        
+        if self.num_used_hadron_types is None:
+            self.num_used_hadron_types = len(hadron_pids)
+        
+        with open(self.pids_map_path, "wb") as f:
+            pickle.dump(self.pids_to_ix, f)
+        print(f"The PID-to-MostCommonID map has been successfully saved in {self.pids_map_path}.")
+
+
+    def _get_n_bins(self, data):
+        """ Compute Scott's normal reference rule and return
+        the number of bins for a histogram plot. """
+        data = np.array(data)
+        std, max_value, min_value = data.std(), data.max(), data.min()
+        data_len = len(data)
+
+        bin_width = 3.49 * std / np.power(2 * data_len, 1/3)
+        n_bins = np.rint((max_value - min_value / bin_width))
+
+        return int(n_bins)
+
+
+    def _plot_dist(self, data, xlabels, ylabels=None, legend_labels=None, labels=None):
+        """ Draw distribution diagrams for a list of three data sets. """
+        plt.clf()
+        _, ax = plt.subplots(1, 3, figsize=(14, 4))
+        colors = ["black", "red", "blue"]
+
+        for i in range(len(data)):
+            samples = data[i]
+            bins = self._get_n_bins(samples)
+            ax[i].hist(samples, bins=bins, color=colors[i], alpha=0.7,
+                       label=legend_labels[i])
+            ax[i].set_xlabel(xlabels[i])
+            if ylabels[i] is not None:
+                ax[i].set_ylabel(ylabels[i])
+            ax[i].legend(loc='upper right')
+
+        plt.tight_layout()
+        filepath = os.path.join(self.data_dir, os.path.normpath("processed/distribution_plots.pdf"))
+        plt.savefig(filepath)
+        print(f"Distribution diagrams have been saved in {filepath}")
+
+
+    def _create_data(self, line, cluster_rest_frame=None, create_dict=False, n_line=None):
+        """ Parse data presented as an event prepared in a specific format.
+        May also be used to parse and return only hadron PIDs needed to create a dictionary. """
+
+        # Splitting clusters   
         items = line.split("|")[:-1]
         clusters = [pd.Series(c.split(";")[:-1]).str.split(',', expand=True) for c in items]
 
+        # Creating a dictionary used for mapping particle IDs to indices
+        if create_dict:
```

Comment:
This section can be moved to `_create_dict` for better readability. Only 2 lines of `_create_data` are shared by code executed when `create_dict==True` and `create_dict==False`

Reply:
Good idea, thanks.